src/lib.rs

```diff
@@ -1,171 +1,164 @@
-    // CAS to see if we won the race or not.  if we lost, free what we
-    // allocated and go home.
-    if SOCKET_FD_INIT != unsafe { SOCKET_FD.compare_and_swap(SOCKET_FD_INIT,
-                                                             socket.as_raw_fd(),
-                                                             Ordering::SeqCst) } {
-        return Ok(())
-    }
-    unsafe {
-        for (i,c) in prefix.bytes().enumerate() {
-            PREFIX[i] = c;
-        }
-        PREFIX_LEN = prefix.len();
-        if PREFIX[PREFIX_LEN-1] != b'.' {
-            PREFIX[PREFIX_LEN] = b'.';
-            PREFIX_LEN += 1;
-        }
-    }
-    socket.into_raw_fd();       // so this doesn't get closed
-    // XXX, that's technically a leaked fd; by design, basically.
-    Ok(())
-}
+const MAX_UDP_PAYLOAD: usize = 576;
 
-#[doc(hidden)]
-pub fn send(strings: &[&[u8]]) {
-    // ideally we'd construct an iov here
-    let fd = unsafe { SOCKET_FD.load(Ordering::Relaxed) };
-    if SOCKET_FD_INIT == fd { return }
-    let mut buf: [u8; 576];
-    buf = unsafe { mem::uninitialized() };
-    let len = {
-        let mut cursor = Cursor::new(&mut buf[..]);
-        let _ = cursor.write_all(unsafe { &PREFIX[..PREFIX_LEN] });
-        for s in strings { let _ = cursor.write_all(s); }
-        cursor.position() as usize
-    };
-    unsafe {
-        libc::send(fd, buf.as_ptr() as *mut c_void, len, MSG_DONTWAIT);
-    }
+pub struct StatsdClient {
+    udp_socket: UdpSocket,
+    prefix: String
 }
 
-use std::cell::RefCell;
+impl StatsdClient {
+    pub fn new(address: &str, prefix: &str) -> Result<StatsdClient, std::io::Error> {
+        let socket = UdpSocket::bind("0.0.0.0:0")?; // NB: CLOEXEC by default
+        socket.set_nonblocking(true)?;
+        socket.connect(address)?;
+        let mut prefix_str = prefix.to_string();
+        prefix_str.push('.');
+        Ok(StatsdClient {
+            udp_socket: socket,
+            prefix: prefix_str
+        })
+    }
 
-#[cfg(target_arch = "x86_64")]
-fn rdtsc() -> u64 {
-    let s: u32;
-    let t: u32;
-    unsafe { asm!("rdtsc" : "={edx}"(s), "={eax}"(t) ::) }
-    ((s as u64) << 32) | (t as u64)
+    pub fn send(&self, strings: &[&str]) {
+        let mut str = String::with_capacity(MAX_UDP_PAYLOAD);
+        str.push_str(&self.prefix);
+        for s in strings { str.push_str(s); }
+        self.udp_socket.send(str.as_bytes());
+    }
 }
 
-fn seed() -> u64 {
-    let seed = 5573589319906701683_u64;
-    let seed = seed.wrapping_mul(6364136223846793005)
-        .wrapping_add(1442695040888963407)
-        .wrapping_add(rdtsc());
-    seed.wrapping_mul(6364136223846793005)
-        .wrapping_add(1442695040888963407)
+#[derive(Debug)]
+pub struct Pcg32Sampling {
+    int_rate: u32,
+    pub buf: String
 }
 
-fn pcg32() -> u32 {
-    thread_local! {
-        static PCG32_STATE: RefCell<u64> = RefCell::new(seed());
+impl Pcg32Sampling {
+    pub fn new(float_rate: f64) -> Box<Pcg32Sampling> {
```

Comment:
It might be useful to (statically?) assert that `float_rate` is between `0.0` and `1.0` (and not `NAN` or `INFINITY`) because casting from a float to an integer can [be undefined behavior](https://github.com/rust-lang/rust/issues/10184).